### PR TITLE
Dev/normalization/windows events

### PR DESCRIPTION
**Pull request by YuvalNaor**
Fixes #

## Proposed Changes

  - adding support for Windows event data sources into the following schemas:
  - Process
  - Authentication
@oshezaf 
@YaronFruchtmann 
  -
  -

### CODE REVIEW

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsSecurityEvent.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ Product:
   Name: Microsoft Windows
 Normalization:
   Schema: Authentication
-  Version: '0.1.0'
+  Version: '1.1.0'
```
Comment (oshezaf) [inline Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsSecurityEvent.yaml:9]: Why? I think that the schema is 0.1.0

File: Parsers/ASimProcessEvent/ProcessEventEmpty.yaml [hunk 2]
```diff
@@ -34,6 +34,8 @@ ParserQuery: |
       EventStartTime:datetime,
       EventEndTime:datetime,
       EventReportUrl:string,
+      EventResult: string,
```
Comment (oshezaf) [inline Parsers/ASimProcessEvent/ProcessEventEmpty.yaml:37]: You will have a merge conflict. I think I also added them.....

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                                              ActorUserId = tostring(Data.SubjectUserSid),
+                                              SubStatus = tostring(Data.SubStatus),
+                                              TargetDomainName = tostring(Data.TargetDomainName),
+                                              TargetSessionId = tostring(Data.TargetLogonId),
+                                              TargetUserId = tostring(Data.TargetUserSid),
+                                              TargetUsername = tostring(Data.TargetUserName),
+                                              SrcDvcHostname = tostring(Data.WorkstationName),
+                                              ActorUserAccount = tostring(Data.SubjectAccount),
+                                              TargetUserAccount = tostring(Data.TargetAccount),
+                                              EventProduct = "Microsoft-Windows-Security-Auditing"
```
Comment (oshezaf) [inline Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml:101]: Should be "Security Events". It is the product as other sources delivering this event. If we want to change, we need to change SecurityEvent as well.

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                                  | extend EventStatus= iff(SubStatus=='0x0',Status,SubStatus)
+                                  // -- creating EventMessage matching EventMessage in SecurityEvent table
+                                  | extend EventMessage = case(EventID == 4634, "4634 - An account was logged off.", 
+                                                              EventID == 4625, "4625 - An account failed to log on.",
+                                                              EventID == 4624 ,"4624 - An account was successfully logged on.",
+                                                              "4647 - User initiated logoff."),
+                                          EventResult = iff(EventID == 4625, 'Failure', 'Success')
+                                  | project-rename 
+                                      TargetDvcHostname = Computer
+                                  // , TargetUserType=AccountType - no AccountType in windowsEvents
```
Comment (oshezaf) [inline Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml:111]: Did you follow with engineering on those issues?
- I recall UID was a bug.
- AccountType: what is the enrichment done in SecurityEvents?

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                                  // , EventOriginalUid = EventOriginId - no EventOriginalId in WindowsEvents
+                                  , EventOriginId=EventID
+                                  | extend  EventCount=int(1)
+                                          , EventSchemaVersion='1.0.0'
```
Comment (oshezaf) [inline Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml:115]: 0.1.0

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                                          , ActorUserIdType='SID'
+                                          , TargetUserIdType='SID'
+                                          , EventVendor='Microsoft'  
+                                          , EventStartTime =TimeGenerated
+                                          , EventEndTime=TimeGenerated
+                                          , EventType=iff(EventOriginId in (LogoffEvents), 'Logoff', 'Logon')
+                                          , ActorUsername = tostring(Data.SubjectUserName)
```
Comment (oshezaf) [inline Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml:122]: Need to be verified. The logic is based on SecurityEvent, which, as we learned, changes things.

File: Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventCreate.yaml [hunk 7]
```diff
@@ -0,0 +1,76 @@
+Parser:
```
Comment (oshezaf) [inline Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventCreate.yaml:1]: We have naming inconsistency. Authentication is WindowsEvents and this is MicrosoftEvents. I think it should be MicrosoftWindowsEvent is the correct one.

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+Normalization:
+  Schema: Authentication
+  Version: '0.1.0'
+References:
+- Title: ASIM Authentication Schema
+  Link: https://aka.ms/AzSentinelAuthenticationDoc
+- Title: ASIM
+  Link: https:/aka.ms/AzSentinelNormalization
+Description: |
+      This Query Parser maps Windows Active Directory sign in logs (WindowsEvent) to the Azure Sentinel Information Model authenticaion schema.
```
Comment (oshezaf) [inline Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml:16]: This is not actually Active Directory related.

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+ParserName: vimAuthenticationWindowsEvent
```
Comment (oshezaf) [inline Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml:17]: MicrosoftWindowsEvent

File: Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventCreate.yaml [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+              let ProcessEvents=(){
+                            WindowsEvent
+                            | where EventID == 4688
+                            // -- Map
+                            | extend
+                              // Event
+                                EventCount = int(1),
+                                EventVendor = 'Microsoft',
+                                EventProduct = 'Windows Events',
+                                EventSchemaVersion = '1.0.0',
```
Comment (oshezaf) [inline Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventCreate.yaml:37]: This and other comments are the same as the other parsers. I may not repeat them all.

File: Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventCreate.yaml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+                                TargetUsernameType = iff (Data.TargetDomainName == '-', 'Simple', 'Windows'),
+                                TargetUserSessionId = tostring(toint(Data.TargetLogonId)),    
+                              // Processes 
+                                ActingProcessId = tostring(toint(Data.ProcessId)),
+                                ActingProcessName = tostring(Data.ParentProcessName),
+                                TargetProcessId = tostring(Data.NewProcessId),
+                                TargetProcessName = tostring(Data.NewProcessName),
+                                TargetProcessCommandLine = tostring(Data.CommandLine),
+                                TargetProcessTokenElevation = tostring(Data.TokenElevationType),
+                                MandatoryLabel = tostring(Data.MandatoryLabel)
```
Comment (oshezaf) [inline Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventCreate.yaml:69]: Do we normalize this?

File: Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventTerminate.yaml [hunk 12]
```diff
@@ -0,0 +1,57 @@
+Parser:
```
Comment (oshezaf) [inline Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventTerminate.yaml:1]: Many comments from other parsers are relevant here.

File: Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventTerminate.yaml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            EventResultDetails = tostring(Data.Status),
+            EventOriginalResultDetails = tostring(Data.Status), 
+          // Device
+            DvcId = tostring(Data.SourceComputerId),
+            DvcHostname = Computer,
+            DvcOs = "Windows",
+          // Users
+            ActorUserId = tostring(Data.SubjectUserSid),
+            ActorUserIdType = "SID",
+            ActorUsername = tostring(Data.SubjectUserName),
```
Comment (oshezaf) [inline Parsers/ASimProcessEvent/ProductParsers/ProcessEventMicrosoftEventTerminate.yaml:43]: The username should include the domain, if available.

**Review by oshezaf — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml (hunk 4 above)
Comment (YuvalNaor) [reply Parsers/ASimAuthentication/ProductParsers/AuthenticationWindowsEvents.yaml:111]: talked about it - will be added later to the original schema

**Review by YuvalNaor — COMMENTED** [review]
finished implementing changes

**Review by oshezaf — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationMicrosoftWindowsEvent.yaml [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  Version: '0.1.0'
+References:
+- Title: ASIM Authentication Schema
+  Link: https://aka.ms/AzSentinelAuthenticationDoc
+- Title: ASIM
+  Link: https:/aka.ms/AzSentinelNormalization
+Description: |
+      This Query Parser maps Windows sign in logs (WindowsEvent) to the Azure Sentinel Information Model authenticaion schema.
+ParserName: vimAuthenticationMicrosoftWindowsEvent
+ParserQuery: |
```
[the remark this answers is not in the document]
Comment (YuvalNaor) [reply Parsers/ASimAuthentication/ProductParsers/AuthenticationMicrosoftWindowsEvent.yaml:18]: for some reason i'm getting YAML errors without those indentations. i'll keep notice

File: Parsers/ASimAuthentication/ProductParsers/AuthenticationMicrosoftWindowsEvent.yaml [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+                                  , EventOriginId=EventID
+                                  | extend  EventCount=int(1)
+                                          , EventSchemaVersion='0.1.0'
+                                          , ActorUserIdType='SID'
+                                          , TargetUserIdType='SID'
+                                          , EventVendor='Microsoft'  
+                                          , EventStartTime =TimeGenerated
+                                          , EventEndTime=TimeGenerated
+                                          , EventType=iff(EventOriginId in (LogoffEvents), 'Logoff', 'Logon')
+                                          , ActorUsername = tostring(Data.SubjectUserName)
```
[the remark this answers is not in the document]
Comment (YuvalNaor) [reply Parsers/ASimAuthentication/ProductParsers/AuthenticationMicrosoftWindowsEvent.yaml:119]: the fields are ordered by dependency list (at you mentioned - when they rely on prior calculation).
changing the comma location

File: Parsers/ASimRegistry/ProductParsers/RegistryEventMicrosoftWindowsEvent.yaml [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+                                    "USER", "HKEY_USERS",   
+                                  ];
+                                let RegistryEvents=(){
+                              WindowsEvent 
+                              | where EventID == 4657
+                              | extend
+                              EventCount = int(1), 
+                              EventVendor = 'Microsoft', 
+                              EventProduct = 'Windows Event', 
+                              EventSchemaVersion = '1.0.0',
```
Comment (oshezaf) [inline Parsers/ASimRegistry/ProductParsers/RegistryEventMicrosoftWindowsEvent.yaml:46]: 0.1.0

**Review by oshezaf — CHANGES_REQUESTED** [review]
Cursery review, but let's go.

File: Parsers/ASimRegistry/ProductParsers/RegistryEventMicrosoftWindowsEvent.yaml (hunk 16 above)
Comment (YuvalNaor) [reply Parsers/ASimRegistry/ProductParsers/RegistryEventMicrosoftWindowsEvent.yaml]: fixed. @oshezaf

**Review by oshezaf — APPROVED** [review]

**Pull request merged**